FloripaPublicTransportation/RestApi.swift

```diff
@@ -0,0 +1,126 @@
+    private static func parseRequest(url: NSURL, params: NSDictionary, taskCompletion: (jsonRows: [NSDictionary]) -> Void)
+    {
+        let request = NSMutableURLRequest(URL: url)
+        request.HTTPMethod = "POST"
+        request.addValue("staging", forHTTPHeaderField: "X-AppGlu-Environment")
+        request.addValue("application/json", forHTTPHeaderField: "Content-Type")
+        
+        let username = "WKD4N7YMA1uiM8V"
+        let password = "DtdTtzMLQlA0hk2C1Yi5pLyVIlAQ68"
+        let credentialsData = "\(username):\(password)".dataUsingEncoding(NSUTF8StringEncoding)
+        let credentialsBase64String = credentialsData?.base64EncodedStringWithOptions([])
+        request.setValue("Basic \(credentialsBase64String!)", forHTTPHeaderField: "Authorization")
+        
+        var parameters: [String: AnyObject] = [:]
+        parameters["params"] = params
+        do {
+            request.HTTPBody = try NSJSONSerialization.dataWithJSONObject(parameters, options: [])
+        }
+        catch let error as NSError {
+            NSLog("NSJSONSerialization.dataWithJSONObject threw an error: \(error.description)")
+            NSLog("Dictionary contents: \(params)")
+        }
+        
+        let session = NSURLSession.sharedSession()
+        let task = session.dataTaskWithRequest(request,
+            completionHandler: { (data, response, error) in
+                guard error == nil else {
+                    NSLog("dataTaskWithRequest sent an error: \(error!.description)")
+                    // calls taskCompletion with an empty NSDictionary:
+                    taskCompletion(jsonRows: [NSDictionary]())
+                    delegate?.onDone()
+                    return
+                }
+                
+                do {
+                    let json = try NSJSONSerialization.JSONObjectWithData(data!, options: .MutableLeaves) as? NSDictionary
+                    if let rows = json!["rows"] as? [NSDictionary] {
+                        // "what to do with the json rows" logic:
+                        taskCompletion(jsonRows: rows)
+                    }
+                    else {
+                        NSLog("Could not interpret \"rows\" as an array of NSDictionary")
+                        NSLog("JSON contents: \(NSString(data: data!, encoding: NSUTF8StringEncoding)!)")
+                    }
+                }
+                catch let error as NSError {
+                    NSLog("NSJSONSerialization.JSONObjectWithData threw an error: \(error.description)")
+                    NSLog("JSON contents: \(NSString(data: data!, encoding: NSUTF8StringEncoding)!)")
+                    // calls taskCompletion with an empty NSDictionary:
+                    taskCompletion(jsonRows: [NSDictionary]())
+                }
+                delegate?.onDone()
+        })
+        
+        task.resume()
+    }
+    
+    static var delegate: ExpectationProtocol? // for tests only
```

Comment:
This is necessary because the request parsing is asynchronous and testing asynchronous methods in Xcode require expectations to be waited and then fulfilled. The use of a delegate allowed to easily define an onDone() function which fulfills that expectation once the completion handler comes to its end.